lib/StackAnalysis/StackAnalysis.cpp

```diff
@@ -383,6 +433,811 @@ void commitToModel(GeneratedCodeBasicInfo &GCBI,
   revng_check(TheBinary.verify(true));
 }
 
+static std::optional<uint64_t> electFSO(const auto &MaybeReturns) {
+  auto It = std::min_element(MaybeReturns.begin(),
+                             MaybeReturns.end(),
+                             [](const auto &LHS, const auto &RHS) {
+                               return LHS.second < RHS.second;
+                             });
+  if (It == MaybeReturns.end())
+    return {};
+  return It->second;
+}
+
+static UpcastablePointer<model::FunctionEdge>
+makeEdge(MetaAddress Destination, model::FunctionEdgeType::Values Type) {
+  model::FunctionEdge *Result = nullptr;
+  if (model::FunctionEdgeType::isCall(Type))
+    Result = new model::CallEdge(Destination, Type);
+  else
+    Result = new model::FunctionEdge(Destination, Type);
+  return UpcastablePointer<model::FunctionEdge>(Result);
+};
+
+template<class FunctionOracle>
+SortedVector<model::BasicBlock>
+CFEPAnalyzer<FunctionOracle>::collectDirectCFG(llvm::BasicBlock *Entry,
+                                               llvm::Function *F) {
+  using namespace llvm;
+  SortedVector<model::BasicBlock> CFG;
+
+  for (auto &BB : *F) {
+    if (GCBI::isJumpTarget(&BB)) {
+      auto Start = getBasicBlockPC(&BB);
+      MetaAddress End;
+      uint64_t Size;
+
+      bool Stop = false;
+      model::BasicBlock Block{ Start };
+      OnceQueue<BasicBlock *> Queue;
+      Queue.insert(&BB);
+
+      while (!Queue.empty()) {
+        BasicBlock *Current = Queue.pop();
+
+        auto [CandidateEnd, CandidateSize] = getPC(Current->getTerminator());
+        if (CandidateEnd > End) {
+          End = CandidateEnd;
+          Size = CandidateSize;
```

Comment:
You don't really need `End` and `Size`. `End` currently is actually `LastInstructionAddress`. I suggest you drop `Size` and embed it into `End`.
If you do so, you don't even need `End` and `Size` but you can act on `Block.End` directly.

Reply:
I understand, this could be simplify. However, I feel like this is kinda more comprehensible / readable, since one may not promptly conclude that `Block.End = End + Size`. Also, updating `Block.End` only once at the very end of the loop seems more reasonable to me than having to update directly this field during the loop. I'd be tempted to say that here, using both `End`/`Size` and be a little bit more verbose is justifiable, and may pay off when coming back on in the future.

Reply:
I don't want to enter long discussions on this.
Let's stick to the facts: `End` is a wrong term, it's not the end of anything. `Size` is also very generic.
You should rename them to `LastInstructionAddress` and `LastInstructionSize` or something like that.
I don't really see a reason to have two variables. If you prefer, you can have `CandidateEnd` and update `Block.End` at ened. But honestly, directly updating `Block.End` seems fine to me.

Reply:
Sure, didn't want to spin this out, and I agree with renaming the variables.

Reply:
Please review if this looks good.